src/ssl.c

```diff
@@ -5173,72 +5174,79 @@ int wolfSSL_Init(void)
+            ret = WC_INIT_E;
         }
 
 #ifdef HAVE_GLOBAL_RNG
-        if (wc_InitMutex(&globalRNGMutex) != 0) {
+        if ((ret == WOLFSSL_SUCCESS) && (wc_InitMutex(&globalRNGMutex) != 0)) {
             WOLFSSL_MSG("Bad Init Mutex rng");
-            return BAD_MUTEX_E;
+            ret = BAD_MUTEX_E;
         }
 #endif
 
 #ifdef OPENSSL_EXTRA
     #ifdef HAVE_ATEXIT
         /* OpenSSL registers cleanup using atexit */
-        if (atexit(AtExitCleanup) != 0) {
+        if ((ret == WOLFSSL_SUCCESS) && (atexit(AtExitCleanup) != 0)) {
             WOLFSSL_MSG("Bad atexit registration");
-            return WC_INIT_E;
+            ret = WC_INIT_E;
         }
     #endif
 
     #ifndef WOLFSSL_NO_OPENSSL_RAND_CB
-        if (wolfSSL_RAND_InitMutex() != 0) {
-            return BAD_MUTEX_E;
+        if ((ret == WOLFSSL_SUCCESS) && (wolfSSL_RAND_InitMutex() != 0)) {
+            ret = BAD_MUTEX_E;
         }
     #endif
-        if (wolfSSL_RAND_seed(NULL, 0) != WOLFSSL_SUCCESS) {
+        if ((ret == WOLFSSL_SUCCESS) &&
+            (wolfSSL_RAND_seed(NULL, 0) != WOLFSSL_SUCCESS)) {
             WOLFSSL_MSG("wolfSSL_RAND_Seed failed");
-            return WC_INIT_E;
+            ret = WC_INIT_E;
         }
 #endif
 
 #ifndef NO_SESSION_CACHE
     #ifdef ENABLE_SESSION_CACHE_ROW_LOCK
         for (i = 0; i < SESSION_ROWS; ++i) {
-            if (wc_InitMutex(&SessionCache[i].row_mutex) != 0) {
+            if ((ret == WOLFSSL_SUCCESS) &&
+                (wc_InitMutex(&SessionCache[i].row_mutex) != 0)) {
                 WOLFSSL_MSG("Bad Init Mutex session");
-                return BAD_MUTEX_E;
+                ret = BAD_MUTEX_E;
             }
         }
     #else
-        if (wc_InitMutex(&session_mutex) != 0) {
+        if ((ret == WOLFSSL_SUCCESS) && (wc_InitMutex(&session_mutex) != 0)) {
             WOLFSSL_MSG("Bad Init Mutex session");
-            return BAD_MUTEX_E;
+            ret = BAD_MUTEX_E;
         }
     #endif
     #ifndef NO_CLIENT_CACHE
-        if (wc_InitMutex(&clisession_mutex) != 0) {
+        if ((ret == WOLFSSL_SUCCESS) &&
+            (wc_InitMutex(&clisession_mutex) != 0)) {
             WOLFSSL_MSG("Bad Init Mutex session");
-            return BAD_MUTEX_E;
+            ret = BAD_MUTEX_E;
         }
     #endif
 #endif
-        if (wc_InitMutex(&count_mutex) != 0) {
+        if ((ret == WOLFSSL_SUCCESS) && (wc_InitMutex(&count_mutex) != 0)) {
             WOLFSSL_MSG("Bad Init Mutex count");
-            return BAD_MUTEX_E;
+            ret = BAD_MUTEX_E;
         }
     }
 
-    if (wc_LockMutex(&count_mutex) != 0) {
+    if ((ret == WOLFSSL_SUCCESS) && (wc_LockMutex(&count_mutex) != 0)) {
         WOLFSSL_MSG("Bad Lock Mutex count");
-        return BAD_MUTEX_E;
+        ret = BAD_MUTEX_E;
     }
 
     initRefCount++;
     wc_UnLockMutex(&count_mutex);
 
-    return WOLFSSL_SUCCESS;
+    if (ret != WOLFSSL_SUCCESS) {
+        (void)wolfSSL_Cleanup(); /* Ignore any error from cleanup */
```

Comment:
This doesn't work right. If a failure occurs before the init above gets called a cleanup might get called before the init. It actually needs to undo each item or track what has been initialized for the cleanup.

Example: `session_mutex` fails. The later `count_mutex` hasn't had init called, but the cleanup routine will try and call `wc_FreeMutex` on it.